File: src/bilibili/broadcast/source.ts
import type { ConfigStore } from "../../config/store";
import type { EventBus } from "../../engine/event-bus";
import { BilibiliLiveSocket, type LiveSocketStatus } from "../live-socket";
import type { BilibiliSource, BilibiliStatus, BroadcastStartInput } from "../types";
import { parseBroadcastGift } from "./parser";
import { fetchDanmuInfo } from "./wbi";

export class BroadcastSource implements BilibiliSource<"broadcast"> {
  readonly type = "broadcast" as const;

  private config: ConfigStore;
  private eventBus: EventBus;
  private socket: BilibiliLiveSocket;
  private roomId: number | null = null;
  private socketStatus: LiveSocketStatus = { connected: false };

  constructor(config: ConfigStore, eventBus: EventBus) {
    this.config = config;
    this.eventBus = eventBus;
    this.socket = new BilibiliLiveSocket();
  }

  async start(input: BroadcastStartInput): Promise<void> {
    if (this.socketStatus.connected) await this.stop();

    const requestedRoomId = input.roomId ?? this.config.bilibili.broadcast.roomId;
    if (!requestedRoomId) throw new Error("roomId required");

    const { key, urls, roomId } = await fetchDanmuInfo(requestedRoomId);
    this.roomId = roomId;

    this.socket.connect({
      label: "Bilibili/Broadcast",
      urls,
      auth: {
        uid: 0,
        roomid: roomId,
        protover: 3,
        platform: "web",
        type: 2,
        key,
      },
      roomId: roomId,
      onMessage: (message) => this.handleMessage(message),
      onStatus: (status) => this.handleSocketStatus(status),
    });

    console.log(`[Bilibili/Broadcast] Started! Room: ${roomId}`);

    await this.config.set({
      bilibili: {
        source: this.type,
        broadcast: { roomId },
      },
    });
  }

  async stop(): Promise<void> {
    this.reset();
  }

  getStatus(): BilibiliStatus {
    return {
      source: this.type,
      connected: this.socketStatus.connected,
      roomId: this.socketStatus.roomId ?? this.roomId ?? undefined,
      error: this.socketStatus.error,
    };
  }

  private handleSocketStatus(status: LiveSocketStatus): void {
    this.socketStatus = status;
    this.eventBus.emit("bilibili:status", this.getStatus());
  }

  private handleMessage(message: unknown): void {
    const gift = parseBroadcastGift(message);
    if (gift) this.eventBus.emit("gift", gift);
  }

  private reset(): void {
    this.roomId = null;
    this.socketStatus = { connected: false };
    this.socket.disconnect();
    this.eventBus.emit("bilibili:status", this.getStatus());
  }
}

File: src/bilibili/broadcast/source.rs
use crate::bilibili::SourceStartResult;
use crate::bilibili::broadcast::parser::parse_broadcast_gift;
use crate::bilibili::broadcast::wbi::fetch_danmu_auth_info;
use crate::bilibili::http_client::HyperHttpClient;
use crate::bilibili::live_socket::{LiveSocketOptions, LiveSocketStatus, run_live_socket};
use crate::config::ConfigHandle;
use crate::config::types::GiftEvent;
use tokio::sync::mpsc;
use tokio_util::sync::CancellationToken;
use tracing::{info, warn};

pub struct BroadcastSource {
    config: ConfigHandle,
    gift_tx: mpsc::Sender<GiftEvent>,
    cancel: CancellationToken,
    http: HyperHttpClient,
}

impl BroadcastSource {
    pub fn new(config: ConfigHandle, gift_tx: mpsc::Sender<GiftEvent>) -> Self {
        Self {
            config,
            gift_tx,
            cancel: CancellationToken::new(),
            http: HyperHttpClient::new(),
        }
    }

    fn reset_cancel(&mut self) -> CancellationToken {
        self.cancel.cancel();
        self.cancel = CancellationToken::new();
        self.cancel.clone()
    }

    pub async fn start(
        &mut self,
        room_id: Option<u64>,
        login_json: Option<String>,
    ) -> Result<SourceStartResult, String> {
        let cancel = self.reset_cancel();

        let cfg = self.config.lock().await;
        let requested = room_id.unwrap_or(cfg.get().bilibili.broadcast.room_id);
        drop(cfg);

        if requested == 0 {
            return Err("roomId required".into());
        }

        let danmu = fetch_danmu_auth_info(&self.http, requested, login_json).await?;

        let gift_tx = self.gift_tx.clone();

        let auth = serde_json::json!({
            "uid": danmu.uid.unwrap_or(0),
            "roomid": danmu.room_id,
            "protover": 3,
            "platform": "web",
            "type": 2,
            "key": danmu.key,
            "buvid": danmu.buvid3,
        });

        let (msg_tx, mut msg_rx) = mpsc::channel::<serde_json::Value>(256);
        let (inner_status_tx, inner_status_rx) = mpsc::channel::<LiveSocketStatus>(16);

        let ls_cancel = cancel.clone();
        tokio::spawn(async move {
            run_live_socket(
                LiveSocketOptions {
                    label: "Bilibili/Broadcast".into(),
                    urls: danmu.urls,
                    auth,
                    room_id: Some(danmu.room_id),
                    on_message: msg_tx,
                    on_status: inner_status_tx,
                },
                ls_cancel,
            )
            .await;
        });

        let gift_tx_h = gift_tx.clone();
        tokio::spawn(async move {
            while let Some(msg) = msg_rx.recv().await {
                if let Some(gift) = parse_broadcast_gift(&msg) {
                    let _ = gift_tx_h.send(gift).await;
                }
            }
        });

        if let Err(e) = self
            .config
            .lock()
            .await
            .update(serde_json::json!({
                "bilibili": {
                    "source": "broadcast",
                    "broadcast": {
                        "roomId": danmu.room_id
                    }
                }
            }))
            .await
        {
            warn!("[Bilibili/Broadcast] Failed to update config: {e}");
        }

        info!("[Bilibili/Broadcast] Started! Room: {}", danmu.room_id);

        Ok(SourceStartResult {
            status_rx: inner_status_rx,
            room_id: Some(danmu.room_id),
            game_id: None,
        })
    }

    pub fn stop(&self) {
        self.cancel.cancel();
    }
}
